File: main/edit-distance/edit-distance.py
class Solution:
    @staticmethod
    def minDistance(word1, word2):
        """
        Wagner-Fischer algorithm
        :type word1: str
        :type word2: str
        :rtype: int
        """
        nrows = 1 + len(word1)
        ncols = 1 + len(word2)
        dist = [[None for _j in range(ncols)] for _i in range(nrows)]

        for j in range(ncols):
            dist[0][j] = j

        for i in range(nrows):
            dist[i][0] = i

        for i, c1 in enumerate(word1):
            for j, c2 in enumerate(word2):
                if c1 == c2:
                    dist[i + 1][j + 1] = dist[i][j]
                else:
                    dist[i + 1][j + 1] = 1 + min(dist[i][j + 1], dist[i + 1][j],
                                                 dist[i][j])

        return dist[-1][-1]

File: main/edit-distance/edit-distance.cs
public class Solution {
    public int MinDistance(string s, string t) {
        if (s.Length < t.Length) (s, t) = (t, s);

        var pre = new int[t.Length + 1];
        var cur = Enumerable.Range(0, pre.Length).ToArray();

        foreach (var i in Enumerable.Range(1, s.Length)) {
            (pre, cur) = (cur, pre);
            cur[0] = i;

            foreach (var j in Enumerable.Range(1, t.Length)) {
                if (s[i - 1] == t[j - 1])
                    cur[j] = pre[j - 1];
                else
                    cur[j] = 1 + Min3(cur[j - 1], pre[j], pre[j - 1]);
            }
        }

        return cur[t.Length];
    }

    private static int Min3(int x, int y, int z)
        => Math.Min(Math.Min(x, y), z);
}
